dhctl/pkg/system/ssh/frontend/reverse-tunnel.go

```diff
@@ -16,68 +16,227 @@ package frontend
 	}
 }
 
 func (t *ReverseTunnel) Up() error {
-	if t.Session == nil {
-		return fmt.Errorf("up tunnel '%s': SSH client is undefined", t.String())
+	_, err := t.upNewTunnel(-1)
+	return err
+}
+
+func (t *ReverseTunnel) upNewTunnel(oldId int) (int, error) {
+	t.tunMutex.Lock()
+	defer t.tunMutex.Unlock()
+
+	if t.started {
+		log.DebugF("[%d] Reverse tunnel already up\n", oldId)
+		return -1, fmt.Errorf("already up")
 	}
 
+	id := rand.Int()
+
+	log.DebugF("[%d] Start reverse tunnel\n", id)
+
 	t.sshCmd = cmd.NewSSH(t.Session).
 		WithArgs(
 			"-N", // no command
 			"-n", // no stdin
 			"-R", t.Address,
 		).
+		WithExitWhenTunnelFailure(true).
 		Cmd()
 
 	err := t.sshCmd.Start()
 	if err != nil {
-		return fmt.Errorf("tunnel up: %w", err)
+		return id, fmt.Errorf("[%d] Cannot start tunnel ssh command: %w", id, err)
 	}
 
 	go func() {
-		err = t.sshCmd.Wait()
+		log.DebugF("[%d] Reverse tunnel started. Wait stop tunnel...\n", id)
+
+		err := t.sshCmd.Wait()
+
+		t.errorCh <- tunnelWaitResult{
+			id:  id,
+			err: err,
+		}
+
+		log.DebugF("[%d] Reverse tunnel was stopped and handled\n", id)
+	}()
+
+	t.started = true
+
+	return id, nil
+}
+
+func (t *ReverseTunnel) isStarted() bool {
+	t.tunMutex.Lock()
+	defer t.tunMutex.Unlock()
+	r := t.started
+	return r
+}
+
+func (t *ReverseTunnel) tryToRestart(id int, killer ReverseTunnelKiller) (int, error) {
+	t.stop(id, false)
+	log.DebugF("[%d] Kill tunnel\n", id)
+	if out, err := killer.KillTunnel(); err != nil {
+		log.DebugF("[%d] Kill tunnel was finished with error: %v; stdout: '%s'\n", id, err, out)
+		return id, err
+	}
+	return t.upNewTunnel(id)
+}
+
+func (t *ReverseTunnel) StartHealthMonitor(checker ReverseTunnelChecker, killer ReverseTunnelKiller) {
+	t.tunMutex.Lock()
+	t.stopCh = make(chan struct{})
+	t.tunMutex.Unlock()
+
+	checkReverseTunnel := func(id int) bool {
+		log.DebugF("[%d] Start Check reverse tunnel\n", id)
+
+		err := retry.NewSilentLoop("Check reverse tunnel", 2, 2*time.Second).Run(func() error {
+			out, err := checker.CheckTunnel()
+			if err != nil {
+				log.DebugF("[%d] Cannot check ssh tunnel: '%v': stderr: '%s'\n", id, err, out)
+				return err
+			}
+
+			return nil
+		})
+
 		if err != nil {
-			f := log.InfoF
-			// signal: killed is normal signal for stopping tunnel process
-			// because we kill it
-			if strings.Contains(err.Error(), "signal: killed") {
-				f = log.DebugF
+			log.DebugF("[%d] Tunnel check timeout, last error: %v\n", id, err)
+			return false
+		}
+
+		log.DebugF("[%d] Tunnel check successful!\n", id)
+		return true
+	}
+
+	go func() {
+		log.InfoLn("Start health monitor")
+		// we need chan for restarting because between restarting we can get stop signal
+		restartCh := make(chan int, 1024)
+		id := -1
+		restart := func(id int) {
+			log.DebugF("[%d] Send restart signal\n", id)
+			restartCh <- id
+			log.DebugF("[%d] Signal was sent. Chan len: %d\n", id, len(restartCh))
+		}
+		for {
+
+			if !checkReverseTunnel(id) {
+				go restart(id)
```

Comment:
This is a potential goroutine leak here, the number of restarts should be limited

Reply:
I think we do no need it here because we do not return error from background process correctly, only panic that is not good user experience. And int another key, we have limited buffer chan

Reply:
Limited channel buffer will not prevent leak as you spawn goroutines prior to blocking on channel operation